homeassistant/components/openuv/__init__.py

```diff
@@ -82,39 +86,99 @@
-            cv.time_period,
-    })
+    DOMAIN:
+        vol.Schema({
+            vol.Required(CONF_API_KEY): cv.string,
+            vol.Optional(CONF_ELEVATION): float,
+            vol.Optional(CONF_LATITUDE): cv.latitude,
+            vol.Optional(CONF_LONGITUDE): cv.longitude,
+            vol.Optional(CONF_BINARY_SENSORS, default={}):
+                BINARY_SENSOR_SCHEMA,
+            vol.Optional(CONF_SENSORS, default={}): SENSOR_SCHEMA,
+            vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+                cv.time_period,
+        })
 }, extra=vol.ALLOW_EXTRA)
 
 
 async def async_setup(hass, config):
     """Set up the OpenUV component."""
-    from pyopenuv import Client
-    from pyopenuv.errors import OpenUvError
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_OPENUV_CLIENT] = {}
+    hass.data[DOMAIN][DATA_OPENUV_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
 
     conf = config[DOMAIN]
-    api_key = conf[CONF_API_KEY]
-    elevation = conf.get(CONF_ELEVATION, hass.config.elevation)
     latitude = conf.get(CONF_LATITUDE, hass.config.latitude)
     longitude = conf.get(CONF_LONGITUDE, hass.config.longitude)
+    elevation = conf.get(CONF_ELEVATION, hass.config.elevation)
+
+    identifier = '{0}, {1}'.format(latitude, longitude)
+
+    if identifier not in configured_instances(hass):
+        hass.async_add_job(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_API_KEY: conf[CONF_API_KEY],
+                    CONF_LATITUDE: latitude,
+                    CONF_LONGITUDE: longitude,
+                    CONF_ELEVATION: elevation
+                }))
+
+    # Store the full OpenUV config for use during entry setup:
+    hass.data[DOMAIN][DATA_OPENUV_CONFIG] = conf
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up OpenUV as config entry."""
+    from pyopenuv import Client
+    from pyopenuv.errors import OpenUvError
+
+    data = hass.data[DOMAIN]
+
+    # Regardless of configuration via config entry or configuration.yaml,
+    # assemble a dict of all config options (with sensible defaults):
+    conf = {
+        **config_entry.data,
+        **data.get(
+            DATA_OPENUV_CONFIG, {
+                CONF_BINARY_SENSORS: {
+                    CONF_MONITORED_CONDITIONS: list(BINARY_SENSORS)
+                },
+                CONF_SENSORS: {
+                    CONF_MONITORED_CONDITIONS: list(SENSORS)
+                }
+            })
+    }
+    conf.setdefault(CONF_ELEVATION, hass.config.elevation)
+    conf.setdefault(CONF_LATITUDE, hass.config.latitude)
+    conf.setdefault(CONF_LONGITUDE, hass.config.longitude)
+
+    # Save the fleshed out config data for use downstream (removing a
+    # user-specified scan interval, since it can't be JSON serialized):
+    config_entry.data = conf
```

Comment:
You are not allowed to change the content of a config entry. Please see https://developers.home-assistant.io/docs/en/creating_component_code_review.html#3-component-platform-communication You can use config_entry.entry_id as a key to link data.